godot-core/src/builtin/quaternion.rs

```diff
@@ -30,10 +30,14 @@ impl Quaternion {
         Self { x, y, z, w }
     }
 
+    /// Creates a Quaternion from a Vector3 and an angle.
+    ///
+    /// # Panics
+    /// If the Vector3 is not normalized.
     pub fn from_angle_axis(axis: Vector3, angle: real) -> Self {
         let d = axis.length();
-        if d == 0.0 {
-            Self::new(0.0, 0.0, 0.0, 0.0)
+        if d != 1.0 {
+            panic!("Attempted to create a Quaternion from a Vector3 that was not normalized.");
```

Comment:
If panic messages are a single sentence, I'd keep them in keyword style and to the point:
```suggestion
            panic!("Quaternion axis {axis:?} is not normalized");
```

Also, `if !cond { panic!(...) }` is better expressed as `assert!(cond, ...)`.

Reply:
Same for below's panic

Reply:
> Also, if !cond { panic!(...) } is better expressed as assert!(cond, ...).

I'm not familiar with panicking from that pattern. Can you point me at docs or an example?

Reply:
I'm just saying that's how assert works -- validating a pre/postcondition.

As for docs, well Rust has docs: https://doc.rust-lang.org/std/macro.assert.html

Reply:
Oh I get what you're getting at. I didn't realize that assert! links to the panic! macro. I always conceived of it from the test context. Will do.

As an aside, constructors are a bit more work. I'm going to punt on adding `from_basis` and `from_axis_axis`. Pushing stuff up later today.